recipes/LJSpeech/TTS/fastspeech2/train.py

```diff
@@ -0,0 +1,369 @@
+        batch: tuple
+            a training batch
+        predictions: tuple
+            predictions (raw output of the FastSpeech2
+             model)
+        """
+        tokens, spectogram, durations, pitch, energy, mel_lengths, input_lengths, labels, wavs = batch
+        mel_post, predict_durations, predict_pitch, predict_energy = predictions
+        self.hparams.progress_sample_logger.remember(
+            target=self.process_mel(spectogram, mel_lengths),
+            output=self.process_mel(mel_post, mel_lengths),
+            raw_batch=self.hparams.progress_sample_logger.get_batch_sample(
+                {
+                    "tokens": tokens,
+                    "input_lengths": input_lengths,
+                    "mel_target": spectogram,
+                    "mel_out": mel_post,
+                    "mel_lengths": mel_lengths,
+                    "durations": durations,
+                    "predict_durations": predict_durations,
+                    "labels": labels,
+                    "wavs": wavs
+
+                }
+            ),
+        )
+
+
+
+
+
+    def process_mel(self, mel, len, index=0):
+        """Converts a mel spectrogram to one that can be saved as an image
+        sample  = sqrt(exp(mel))
+        Arguments
+        ---------
+        mel: torch.Tensor
+            the mel spectrogram (as used in the model)
+        len: int
+            length of the mel spectrogram
+        index: int
+            batch index
+        Returns
+        -------
+        mel: torch.Tensor
+            the spectrogram, for image saving purposes
+        """
+        assert mel.dim() == 3
+        return torch.sqrt(torch.exp(mel[index][:len[index]]))
+
+
+    def on_stage_end(self, stage, stage_loss, epoch):
+        """Gets called at the end of an epoch.
+        Arguments
+        ---------
+        stage : sb.Stage
+            One of sb.Stage.TRAIN, sb.Stage.VALID, sb.Stage.TEST
+        stage_loss : float
+            The average loss for all of the data processed in this stage.
+        epoch : int
+            The currently-starting epoch. This is passed
+            `None` during the test stage.
+        """
+        # Store the train loss until the validation stage.
+        if stage == sb.Stage.TRAIN:
+            self.train_loss = stage_loss
+        # Summarize the statistics from the stage for record-keeping.
+        else:
+            stats = {
+                "loss": stage_loss,
+            }
+
+        # At the end of validation, we can write
+        if stage == sb.Stage.VALID:
+            # Update learning rate
+            self.last_epoch = epoch
+            lr = self.optimizer.param_groups[-1]["lr"]
+
+            # The train_logger writes a summary to stdout and to the logfile.
+            self.hparams.train_logger.log_stats(  # 1#2#
+                stats_meta={"Epoch": epoch, "lr": lr},
+                train_stats={"loss": self.train_loss},
+                valid_stats=stats,
+            )
```

Comment:
It can be useful to show the components of the loss as other recipes did.(mel_loss:..., dur_loss:..., pitch_loss:..., energy_loss:...)